Versioning.md

```diff
@@ -0,0 +1,83 @@
+# Versioning
+
+This document describes the versioning policy for this repository.
+
+## Goals
+
+### API Stability
+
+Once the API for a given signal (spans, logs, metrics, baggage) has been officially released, that API module will function with any SDK that has the same major version, and equal or greater minor or patch version.
+
+For example, libraries that are instrumented with `opentelemetry 1.0.1` will
+function in applications using `opentelemetry 1.11.33` or `opentelemetry
+1.3.4`.
+
+### SDK Stability
+
+Public portions of the SDK (constructors, configuration, end-user interfaces)
+must remain backwards compatible. Internal types are allowed to break.
+
+## Policy
+
+* Release versions will follow [SemVer 2.0](https://semver.org/).
+* Only single source package containing api and sdk for all signals would be released as part of the new GitHub release.
+* Experimental releases: New (unstable) telemetry signals and features will be introduced behind feature flag using C macro define.
+  ```
+    #ifdef FEATURE_FLAG
+         <metrics api/sdk definitions>
+    #endif
+  ```
+
+  As we deliver package in source form, and the user is responsible to build them for their platform, they must be
+  aware of these feature flags (documented in CHANGELOG.md), and enable it explicitly through their build system (cmake
+  , bazel or others ) to use the preview feature.
+
+  The guidelines in creating feature flag would be:
+  - Naming:
+
+        - `ENABLE_<SIGNAL>_PREVIEW` : For experimetal release of signal api/sdks eg, METRICS_PREVIEW, LOGGING_PREVIEW,
+
+        - `ENABLE_<SIGNAL>_<FEATURE_NAME>_PREVIEW` : For experimental release for any feature within stable signal. Eg, TRACING_JAEGER_PREVIEW to release the experimental Jaeger exporter for tracing.
+
+  - Cleanup: It is good practice to keep feature-flags shortlived as possible. And, aslo important to keep the number of them low. They should be used such that it is easy to remove/cleanup them once the experimental feature is stable.
+
+
+* New signals will be stabilized via a **minor version bump**, and are not allowed to break existing stable interfaces.
+Feature flag would be removed once we have a stable implementation for the signal.
+
+* GitHub releases will be made for all released versions.
+
+## Example Versioning Lifecycle
+
+Purely for illustration purposes, not intended to represent actual releases:
```

Comment:
I assume we are either adding git tags on `master` or create `release/v1.0.0` branch.

For API+SDK in this repo, should we assume that we attempt to keep both API+SDK in sync at all times?

For release packaging: assuming we use either CMake (git+tag) or vcpkg or bazel, isn't our mainline release always produce just one artifact, that is:
- API headers (header-only?), plus
- SDK implementation combined

And we do not release separate packages per se? Practical example: if I install a package on Ubuntu, usually I have `package` - the runtime, and `package-dev` - headers and libraries, documentation, possibly source, etc..

I understand the merit of logically splitting API vs. SDK, so that a vendor can implement their own SDK and exporters based on ABI-stable API. But from end-user consumption perspective, what are the release packages that we actually deliver and install?

Example:

```console
vcpkg install opentelemetry
```
- that'd be ideal.

OR

```console
apt install opentelemetry
apt install opentelemetry-dev
```
That's how I would've done the runtime vs. lib+headers+src+doc 'developer` package.

Best how we can accommodate the API+SDK is to combine them into one ship'able artifact, i.e. one source tarball with given version v1.0.0. What are your own thoughts on that? i.e. irrespective of what OTEP is suggesting.

Reply:
My thoughts are also on similar lines. Ensuring that the API + SDK are sync all the time, and deliver one source package with both api + sdk as single version. Even if we want to package it through some package manager, let it be single package containing both headers (api) and lib (sdk).

Reply:
Yes, I think OTEP says should and may, and under the may - it recommends producing one package. It would be easier for us if we always enforce one common tag, common baseline, for api + sdk. Every time we change something either in API or in SDK, we update the tag, so the next release bumps-up both at the same time.

Reply:
..for C++, I believe, the bigger challenge is that since we do not provide ABI stability guarantee for exporters -- it means vendors must build their own entire SDK with their custom exporter. Maybe we need to do the write-up later on how vendors take the `opentelemetry` package `v1.0.0` , and produce their own build of it as `opentelemetry-sdk-vendor` package `v1.0.0` that contains their own proprietary exporter implementation. Correspondingly, it'd make sense to later on describe how a sample app would be instrumented with either vendor SDK package, and maybe even with two different vendor packages at the same time, if necessary, i.e. using TraceProvider from Vendor A and TraceProvider from Vendor B.

Reply:
@maxgolov  Based on the recent comment here 
https://github.com/open-telemetry/oteps/pull/143#discussion_r541305162

Seems decision is moving towards having different versions of api and sdk, and ask is for us to get aligned to this.

Reply:
I think the idea there is the API is less-able-to-change than the SDK.  Also, there's the notion that someone could take a dependency on the API itself without the SDK.


For C++, a few guiding questions:
- Would we allow, say, a version 2.0 of the SDK to still have the API namespace v1 with compatibility?
- Do we expect projects to take a dependency on just the API without the SDK in the C++ world?
- Do we want a unified notion of artifact naming + versioning for various ecosystems?
   vcpkg, deb, rpm, source, bazel.
   Does this mean we'd own each of these packaging scenarios?

Reply:
> Would we allow, say, a version 2.0 of the SDK to still have the API namespace v1 with compatibility?
As per what we plan to follow,  API module will function with any SDK that has the same MAJOR version and equal or greater MINOR or PATCH version. As per otps, there is not going to be any 2.0 release for API and SDK.

> Do we expect projects to take a dependency on just the API without the SDK in the C++ world?
The `opentelemetry-api` should be enough for instrumenting any library, with final choice of SDK deferred to application developer. 

> Do we want a unified notion of artifact naming + versioning for various ecosystems?
We don't plan to own all the ecosystems, the official release would be source(zip/tar) package(s): 
Either single source package like 
   `opentelemtry-1.0.0`   
or several:
   `opentelemetry-api-1.0.0`, `opentelemetry-sdk-1.0.1`, `opentelemetry-exporter-zipkin-1.0.1`